client/client.go

```diff
@@ -326,66 +426,27 @@ func decodeInto(reader io.Reader, v interface{}) error {
 	return nil
 }
 
-// doSync performs a request to the given path using the specified HTTP method.
-// It expects a "sync" response from the API and on success decodes the JSON
-// response payload into the given value using the "UseNumber" json decoding
-// which produces json.Numbers instead of float64 types for numbers.
-func (client *Client) doSync(method, path string, query url.Values, headers map[string]string, body io.Reader, v interface{}) (*ResultInfo, error) {
-	var rsp response
-	if err := client.do(method, path, query, headers, body, &rsp); err != nil {
-		return nil, err
-	}
-	if err := rsp.err(client); err != nil {
-		return nil, err
-	}
-	if rsp.Type != "sync" {
-		return nil, fmt.Errorf("expected sync response, got %q", rsp.Type)
-	}
-
-	if v != nil {
-		if err := decodeWithNumber(bytes.NewReader(rsp.Result), v); err != nil {
-			return nil, fmt.Errorf("cannot unmarshal: %w", err)
-		}
-	}
-
-	client.warningCount = rsp.WarningCount
-	client.warningTimestamp = rsp.WarningTimestamp
-
-	return &rsp.ResultInfo, nil
+func (client *Client) doSync(method, path string, query url.Values, headers map[string]string, body io.Reader, v interface{}) (*RequestResponse, error) {
+	return client.Requester.Do(context.Background(), &RequestOptions{
+		Method:  method,
+		Path:    path,
+		Query:   query,
+		Headers: headers,
+		Body:    body,
+	}, v)
 }
 
-func (client *Client) doAsync(method, path string, query url.Values, headers map[string]string, body io.Reader) (changeID string, err error) {
-	_, changeID, err = client.doAsyncFull(method, path, query, headers, body)
-	return
-}
-
-func (client *Client) doAsyncFull(method, path string, query url.Values, headers map[string]string, body io.Reader) (result json.RawMessage, changeID string, err error) {
-	var rsp response
-
-	if err := client.do(method, path, query, headers, body, &rsp); err != nil {
-		return nil, "", err
-	}
-	if err := rsp.err(client); err != nil {
-		return nil, "", err
-	}
-	if rsp.Type != "async" {
-		return nil, "", fmt.Errorf("expected async response for %q on %q, got %q", method, path, rsp.Type)
-	}
-	if rsp.StatusCode != 202 {
-		return nil, "", fmt.Errorf("operation not accepted")
-	}
-	if rsp.Change == "" {
-		return nil, "", fmt.Errorf("async response without change reference")
-	}
-
-	return rsp.Result, rsp.Change, nil
+func (client *Client) doAsync(method, path string, query url.Values, headers map[string]string, body io.Reader, v interface{}) (*RequestResponse, error) {
+	return client.Requester.Do(context.Background(), &RequestOptions{
+		Method:  method,
+		Path:    path,
+		Query:   query,
+		Headers: headers,
+		Body:    body,
+		Async:   true,
+	}, v)
 }
 
-// ResultInfo is empty for now, but this is the mechanism that conveys
-// general information that makes sense to requests at a more general
-// level, and might be disconnected from the specific request at hand.
-type ResultInfo struct{}
```

Comment:
This is effectively renamed to ```RequestResponse```. We could reuse the name if it is really an issue, but the new naming feels consistent with the new interface.